File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
public:
void dfs(int r,int c,vector<vector<int>>&visited,vector<vector<char>>& board){
  visited[r][c] = 1;
 int nrow[] = {0,0,-1,+1};
 int ncol[] = {-1,+1,0,0};
  for(int i=0;i<4;i++){
    int row = r + nrow[i];
    int col = c + ncol[i];
    if(row>=0 && row < board.size() && col >= 0 && col < board[0].size() && board[row][col] == 'O' && visited[row][col] == 0){
       visited[row][col] = 1;
       dfs(row,col,visited,board);
    }

  }
}
    void solve(vector<vector<char>>& board) {
        vector<vector<int>>visited(board.size(),vector<int>(board[0].size(),0));
        queue<pair<int,int>>q;

        int m = board.size()-1;
        for(int i=0;i<board.size();i++){
         for(int j=0;j<board[i].size();j++){
            if( i == 0 && board[i][j] == 'O' && visited[i][j] == 0){
              dfs(i,j,visited,board);
            }
            if(i == board.size()-1 && board[i][j] == 'O' && visited[i][j] == 0){
                  dfs(i,j,visited,board);
            }
             if(j == 0 && board[i][j] == 'O' && visited[i][j] == 0){
                  dfs(i,j,visited,board);
            }
            if(j == board[i].size()-1 && board[i][j] == 'O' && visited[i][j] == 0){
                  dfs(i,j,visited,board);
            }
         }
        }
       
        int row[] ={0,0,+1,-1};
     int col[] = {-1,+1,0,0}; 
       
       for(int i=0;i<board.size();i++){
        for(int j=0;j<board[i].size();j++){
          if(board[i][j] == 'O' && visited[i][j] == 0){
             board[i][j] = 'X';
          }
        }
       }
       
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
     class Pair{
        int row; int col;
        Pair(int row, int col ){
            this.row = row;
            this.col = col;
            
        }
    }
    public void solve(char[][] mat) {
         Queue<Pair> q = new LinkedList<>();
         
          int[][] ans = new int[mat.length][mat[0].length];
           int[][] vis = new int[mat.length][mat[0].length];;
         int n = mat.length-1;
         
          for(int[] a : vis){
           Arrays.fill(a,0);
          }
          for(int i=0;i<mat.length;i++){
           
               if(mat[i][0] == 'O'){
                q.add(new Pair(i,0));
               }
               if(mat[i][mat[0].length-1] == 'O'){
                q.add(new Pair(i,mat[0].length-1));
               }
            
          }
           for(int i=0;i<mat[0].length;i++){
           
               if(mat[0][i] == 'O'){
                q.add(new Pair(0,i));
               }
               if(mat[n][i] == 'O'){
                q.add(new Pair(n,i));
               }
            
          }
             int rows[] = {0,0,-1,+1};
           int cols[]= {-1,+1,0,0};
           while(q.size() > 0){
             Pair p = q.poll();
             vis[p.row][p.col] = 1;
             for(int i=0;i<4;i++){
               int r = p.row + rows[i];
               int c = p.col + cols[i];

               if(r>=0 && r < mat.length && c>=0 && c < mat[0].length && mat[r][c] == 'O' && vis[r][c] == 0){
                  q.add(new Pair(r,c));
                  vis[r][c] = 1;
                 
               }
             }

          }

          for(int i=0;i<mat.length;i++){
            for(int j=0;j<mat[0].length;j++){
                if(mat[i][j] == 'O' && vis[i][j] == 0){
                mat[i][j] = 'X';
               }
            }
          }
         


           
    }
}